// shopify-theme-tools/dist/writers/ShopifyAssetCreator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
class ShopifyAssetCreator {
    constructor(prefix, sectionName) {
        this.processDir = process.cwd();
        this.prefix = prefix;
        this.sectionName = sectionName;
    }
    createCss() {
        if (!fs_1.default.existsSync(path_1.default.join(this.processDir, "assets"))) {
            throw new Error("Missing assets directory");
        }
        const destinationPath = path_1.default.join(this.processDir, "assets", this.prefix + "-" + this.sectionName + ".css");
        if (!fs_1.default.existsSync(destinationPath)) {
            fs_1.default.writeFileSync(destinationPath, `.${this.prefix}-${this.sectionName} {\n  /* Your style here */\n}`);
        }
    }
    createJs() {
        if (!fs_1.default.existsSync(path_1.default.join(this.processDir, "assets"))) {
            throw new Error("Missing assets directory");
        }
        const destinationPath = path_1.default.join(this.processDir, "assets", this.prefix + "-" + this.sectionName + ".js");
        if (!fs_1.default.existsSync(destinationPath)) {
            fs_1.default.writeFileSync(destinationPath, "");
        }
    }
}
exports.default = ShopifyAssetCreator;
