// shopify-theme-tools/dist/clients/libreTranslateClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.libreTranslateClient = void 0;
const settings_1 = require("../settings");
const url = settings_1.settings.translationServiceUrl;
const apiKey = settings_1.settings.translationServiceApiKey;
const defaultLanguage = settings_1.settings.defaultLanguage;
const supportedLanguages = settings_1.settings.supportedLanguages;
let _isAvailable = true;
const translate = async (text, targetLanguage) => {
    if (!supportedLanguages.includes(targetLanguage)) {
        throw new Error(`Language ${targetLanguage} is not supported`);
    }
    if (!_isAvailable) {
        return text;
    }
    try {
        const res = await fetch(`${url}/translate`, {
            method: "POST",
            body: JSON.stringify({
                q: text,
                source: defaultLanguage,
                target: targetLanguage,
                format: "text",
                alternatives: 0,
                api_key: apiKey,
            }),
            headers: { "Content-Type": "application/json" },
        });
        const data = await res.json();
        _isAvailable = true;
        return data.translatedText;
    }
    catch (e) {
        _isAvailable = false;
        return "";
    }
};
const isAvailable = async () => {
    await translate("", defaultLanguage);
    return _isAvailable;
};
exports.libreTranslateClient = {
    translate,
    isAvailable,
};
